### ISSUE TITLE
Using MLFlow>=2.18.0 with ThreadRunner creates several runs

### ISSUE THREAD

**Issue by kacper-ki** [Galileo-Galilei/kedro-mlflow#613]
## Description
Updating to MLFlow 2.18.0 makes kedro-mlflow's datasets create and log to different runs (new run per artifact). It also breaks kedro-mlflow's test suite.

## Context
Cannot use MLFlow >= 2.18

## Steps to Reproduce

1. Log artifact(s) with any kedro-mlflow's dataset

## Expected Result

Artifact should be logged to the initial run

## Actual Result

Artifact is logged to a new run

Test suite failing on 2.18
```
===short test summary info ===
FAILED tests/mlflow/test_kedro_pipeline_model.py::test_kedro_pipeline_model_save_and_load[pipeline_inference_with_parameters-catalog_with_parameters-data-result3] - mlflow.exceptions.MlflowException: The provided MLFLOW_EXPERIMENT_ID environment variable value `514828615563594097` does not exist in the tracking server. Provide a valid experiment_id.
...

78 failed, 165 passed, 2 skipped, 1 xfailed, 41 warnings in 74.11s (0:01:14)
```

Initial run and subsequent runs with artifacts logged

<img width="1102" alt="image" src="https://github.com/user-attachments/assets/708556ca-b228-4a5d-aa9e-a544a4ecac6c">

## Your Environment

* `kedro` and `kedro-mlflow` version used (`pip show kedro` and `pip show kedro-mlflow`): 0.19.8 and 0.13.3
* Python version used (`python -V`): 3.10.14
* Operating system and version: MacOS M2 15.1.1

## Does the bug also happen with the last version on master?

Yes

**Comment by Calychas** [comment]
I forgot to add - 2.17.2 works correctly

**Comment by Galileo-Galilei** [comment]
This is very confusing. I've spent a lot of time trying to reproduce the error, and I can't get it properly. 

⚠️  **Failing tests**: I can see the tests are failing, and  if I run the full tests suite I can get all these errors locally. 
❓ However when I run the test suite partially (with ``pytest -k <regex>``) never manage to get a failing test. Running tests locally in a notebook works perfectly. I am trying to find a way to reproduce the bug. 

⚠️  **Artifact logging in the wrong run**: Fact is that you are experiencing issues
❓ However when I run kedro mlflow on the default spaceflights pandas, I can't reproduce it. Do you use a specific Runner (like ParallelRunner or ThreadRunner?) ? Do you have a reproducible example to share?

**Comment by Galileo-Galilei** [comment]
The root of all evil is https://github.com/mlflow/mlflow/pull/13456. 

They decided to make mlflow threadsafe by resetting env var at the thread level. This causes the differents bugs (I guess the one in the tests is much more subtle because it depends on the running order, I'll get back to it).  Unfortunately, this is something I can hardly change on my own : mlflow forces each thread to be a different run.

I will still look for a solution, but it can be anything other than hacky. I think the best solution here is to open an issue over mlflow's repo, and explain what is at stake for us but I doubt this will change anything.

**Comment by Calychas** [comment]
Hey! Thanks for looking into this!
Is it also possible that this is the root issue https://github.com/mlflow/mlflow/releases/tag/v2.19.0#:~:text=%5BTracking%5D%20Fix%20mlflow.end_run%20inside%20a%20MLflow%20run%20context%20manager%20(%2313888%2C%20%40WeichenXu123) ? Did you try on 2.19.0?

**Comment by Galileo-Galilei** [comment]
Yes I tried with 0.19 and this persist. I've made extensive tests and the bug is introduced in this very commit : 

![image](https://github.com/user-attachments/assets/e00d7d53-1023-484d-8b1d-1d550573fba6)

Cloning the repo and running tests on spaceflights padnas does work with the commit before (81bb...), and does fail with above commit (487db...)+ the code is very explicit : the global variable having the active run stack is cleared inside each thread so we now have an active_run_stack at the thread level. I've fixed all tests (see the associated MR), but your problem is another issue not covered by a test. I am still trying to get a working solution.
